flink-connector-dynamodb/src/main/java/org/apache/flink/streaming/connectors/dynamodb/sink/DynamoDbSinkWriter.java

```diff
@@ -0,0 +1,307 @@
+                    CONDITIONAL_CHECK_FAILED_EXCEPTION_CLASSIFIER,
+                    VALIDATION_EXCEPTION_CLASSIFIER,
+                    getSdkClientMisconfiguredExceptionClassifier());
+
+    /* A counter for the total number of records that have encountered an error during put */
+    private final Counter numRecordsSendErrorsCounter;
+
+    /* A counter for the total number of records that were returned by DynamoDB as unprocessed and were retried */
+    private final Counter numRecordsReturnedAsUnprocessedCounter;
+
+    /* The sink writer metric group */
+    private final SinkWriterMetricGroup metrics;
+
+    /* The asynchronous http client for the asynchronous DynamoDb client */
+    private final SdkAsyncHttpClient httpClient;
+    private final DynamoDbAsyncClient client;
+    private final boolean failOnError;
+    private final String tableName;
+
+    private final List<String> overwriteByPartitionKeys;
+
+    public DynamoDbSinkWriter(
+            ElementConverter<InputT, DynamoDbWriteRequest> elementConverter,
+            InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes,
+            boolean failOnError,
+            String tableName,
+            List<String> overwriteByPartitionKeys,
+            Properties dynamoDbClientProperties,
+            Collection<BufferedRequestState<DynamoDbWriteRequest>> states) {
+        super(
+                elementConverter,
+                context,
+                maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBatchSizeInBytes,
+                maxTimeInBufferMS,
+                maxRecordSizeInBytes,
+                states);
+        this.failOnError = failOnError;
+        this.tableName = tableName;
+        this.overwriteByPartitionKeys = overwriteByPartitionKeys;
+        this.metrics = context.metricGroup();
+        this.numRecordsSendErrorsCounter = metrics.getNumRecordsSendErrorsCounter();
+        this.numRecordsReturnedAsUnprocessedCounter =
+                metrics.counter("numRecordsReturnedAsUnprocessedCounter");
+        this.httpClient = AWSGeneralUtil.createAsyncHttpClient(dynamoDbClientProperties);
+        this.client = buildClient(dynamoDbClientProperties, httpClient);
+    }
+
+    private DynamoDbAsyncClient buildClient(
+            Properties dynamoDbClientProperties, SdkAsyncHttpClient httpClient) {
+        AWSGeneralUtil.validateAwsCredentials(dynamoDbClientProperties);
+
+        return AWSAsyncSinkUtil.createAwsAsyncClient(
+                dynamoDbClientProperties,
+                httpClient,
+                DynamoDbAsyncClient.builder(),
+                DynamoDbConfigConstants.BASE_DYNAMODB_USER_AGENT_PREFIX_FORMAT,
+                DynamoDbConfigConstants.DYNAMODB_CLIENT_USER_AGENT_PREFIX);
+    }
```

Comment:
```suggestion
    public DynamoDbSinkWriter(
            ElementConverter<InputT, DynamoDbWriteRequest> elementConverter,
            InitContext context,
            int maxBatchSize,
            int maxInFlightRequests,
            int maxBufferedRequests,
            long maxBatchSizeInBytes,
            long maxTimeInBufferMS,
            long maxRecordSizeInBytes,
            boolean failOnError,
            String tableName,
            List<String> overwriteByPartitionKeys,
            Properties dynamoDbClientProperties,
            Collection<BufferedRequestState<DynamoDbWriteRequest>> states) {
        super(
                elementConverter,
                context,
                maxBatchSize,
                maxInFlightRequests,
                maxBufferedRequests,
                maxBatchSizeInBytes,
                maxTimeInBufferMS,
                maxRecordSizeInBytes,
                states);
        this.failOnError = failOnError;
        this.tableName = tableName;
        this.overwriteByPartitionKeys = overwriteByPartitionKeys;
        this.metrics = context.metricGroup();
        this.numRecordsSendErrorsCounter = metrics.getNumRecordsSendErrorsCounter();
        this.numRecordsReturnedAsUnprocessedCounter =
                metrics.counter("numRecordsReturnedAsUnprocessedCounter");
        this.httpClient = AWSGeneralUtil.createAsyncHttpClient(overrideClientProperties(dynamoDbClientProperties));
        this.client = buildClient(dynamoDbClientProperties, httpClient);
    }
    
    
    private Properties overrideClientProperties(Properties dynamoDbClientProperties) {
        Properties overridenProperties = new Properties();
        overridenProperties.putAll(dynamoDbClientProperties);

        // Specify HTTP1_1 protocol since DynamoDB endpoint doesn't support HTTP2
        overridenProperties.put(AWSConfigConstants.HTTP_PROTOCOL_VERSION, "HTTP1_1");
        return overridenProperties;
    }

    private DynamoDbAsyncClient buildClient(
            Properties dynamoDbClientProperties, SdkAsyncHttpClient httpClient) {
        AWSGeneralUtil.validateAwsCredentials(dynamoDbClientProperties);

        return AWSAsyncSinkUtil.createAwsAsyncClient(
                dynamoDbClientProperties,
                httpClient,
                DynamoDbAsyncClient.builder(),
                DynamoDbConfigConstants.BASE_DYNAMODB_USER_AGENT_PREFIX_FORMAT,
                DynamoDbConfigConstants.DYNAMODB_CLIENT_USER_AGENT_PREFIX);
    }
```